File: ocaml/src/p0071.ml
(* Project Euler: Problem 71 *)

(*
 * Farey sequence
 *
 * 2/5, 3/7
 *   -> 2/5, (2+3)/(5+7), 3/7
 *   -> 2/5, (2+3)/(5+7), (2+3+3)/(5+7+7), 3/7
 *   -> 2/5, (2+3)/(5+7), (2+3+3)/(5+7+7), (2+3+3+3)/(5+7+7+7), 3/7
 *    ...
 *   -> 2/5, ..., (2+3x)/(5+7x), 3/7
 *
 *     5+7x <= 1_000_000
 *)

open Core

let compute n =
  let x = (n - 5) / 7 in
  2 + (3 * x)
;;

let solve () = compute 1_000_000 |> Int.to_string

(* Test *)

let%test_unit "8" = [%test_eq: int] (compute 8) 2
let%test_unit "1_000_000" = [%test_eq: int] (compute 1_000_000) 428570

File: rust/src/bin/p0071.rs
// Project Euler: Problem 71

/*
  Farey sequence

  2/5, 3/7
    -> 2/5, (2+3)/(5+7), 3/7
    -> 2/5, (2+3)/(5+7), (2+3+3)/(5+7+7), 3/7
    -> 2/5, (2+3)/(5+7), (2+3+3)/(5+7+7), (2+3+3+3)/(5+7+7+7), 3/7
     ...
    -> 2/5, ..., (2+3x)/(5+7x), 3/7

      5+7x <= 1_000_000
*/

euler::run_solver!(71);

fn solve() -> String {
    compute(1_000_000).to_string()
}

fn compute(limit: i64) -> i64 {
    debug_assert!(limit >= 8);

    2 + 3 * ((limit - 5) / 7)
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0071_8() {
        assert_eq!(compute(8), 2);
    }
    #[test]
    fn p0071_1000000() {
        assert_eq!(compute(1_000_000), 428570);
    }
}
